akka-actor-typed/src/main/scala/akka/actor/typed/internal/ActorContextImpl.scala

```diff
@@ -67,5 +72,35 @@ import scala.concurrent.ExecutionContextExecutor
    * Otherwise "ambiguous reference to overloaded definition" because Function is lambda.
    */
   @InternalApi private[akka] def internalSpawnAdapter[U](f: U ⇒ T, _name: String): ActorRef[U]
+
+  // Scala impl
+  override def ask[Req, Res](otherActor: ActorRef[Req], createMessage: ActorRef[Res] ⇒ Req)(responseToOwnProtocol: Try[Res] ⇒ T)(implicit timeout: Timeout, classTag: ClassTag[Res]): Unit = {
+
+    import akka.actor.typed.scaladsl.AskPattern._
+
+    implicit val scheduler = system.scheduler
+
+    (otherActor ? createMessage)
+      .mapTo[Res]
+      .onComplete(t ⇒ self ! responseToOwnProtocol(t))
```

Comment:
I think it would be interesting to explore if we can run the message adapter function in the actor instead of in another EC. Then it would be fool-proof, and it would even be possible to access mutable state of the actor in the function. 

Also, less surprising with regards to message order because the replied is processed immediately.
E.g. two ask requests are sent after each other to the same destination actor, which replies in order. But this implementation doesn't guarantee that the replies are received in order, because of the `onComplete`.

To completely replace pipeTo we could have a variant of `ask` that also takes a `replyTo: ActorRef` parameter which can be used if the final reply should go to a third actor. The message mapping would still be done in the actor initiating the ask.

Reply:
After thinking some more and some failed experiments my conclusion is that we can't fix the ordering thing, since in the end we must use ordinary `?` to be able to capture context (the message mapping function). That means different actor refs for the replies and the ordering is not guaranteed (and Artery has different lanes). spawnAdapter would also be different actor refs.

One thing we could make more safe is to not apply the message mapping function here (in other thread) but send an internal `AskResponse` message to self. It would carry the response and the mapping function. It would be understood by `ActorAdapter` and the function is applied there (in `receive`) before delegating to the `Behavior`.

See also https://github.com/akka/akka/pull/24305